File: 8-string-to-integer-atoi/8.py
class Solution:
    def myAtoi(self, s: str) -> int:
        ret = 0
        numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
        neg = False
        start = 0
        while start < len(s):
            if s[start] == ' ':
                start += 1
            else:
                break

        if start == len(s):
            return 0
        if s[start] not in numbers and s[start] != '-' and s[start] != '+':
            return 0

        changed = False
        num_start = False

        for i in range(start, len(s)):
            if s[i] == '-':
                if changed or num_start:
                    break
                else:
                    changed = True
                neg = True
                continue
            elif s[i] == '+':
                if changed or num_start:
                    break
                else:
                    changed = True
                neg = False
                continue
            elif s[i] not in numbers:
                break
            num_start = True
            ret *= 10
            ret += int(s[i])

        if neg:
            if ret >= 2147483648:
                return -2147483648
            else:
                return -ret
        else:
            if ret >= 2147483647:
                return 2147483647
            else:
                return ret

File: 8-string-to-integer-atoi/8.cpp
class Solution {
public:
    int myAtoi(string s) {
        string str;
        uint64_t num = 0;
        int i = 0;
        bool neg = false;
        while (i < s.size()) {
            if (s[i] == ' ') {
                ++i;
                continue;
            } else {
                break;
            }
        }
        if (s[i] == '+') {
            neg = false;
            ++i;
        } else if (s[i] == '-') {
            neg = true;
            ++i;
        }
        while (i < s.size()) {
            if ((s[i] - '0') >= 0 && (s[i] - '0') <= 9) {
                if (num >= 2147483648)
                    break;
                num = num * 10 + s[i] - '0';
                ++i;
            } else {
                break;
            }
        }
        if (!neg && num >= 2147483647) {
            return 2147483647;
        } else if (neg && num >= 2147483648) {
            return -2147483648;
        }
        if (neg)
            return -num;
        return num;
    }
};
